### ISSUE TITLE
Clouddriver failed to clean up gitrepo when contains symlink

### ISSUE THREAD

**Issue by weitie-okta** [spinnaker/spinnaker#6759]
### Issue Summary:

Clouddriver failed to clean up local gitrepo when it contains symlink

The issue has been reported before at https://github.com/spinnaker/spinnaker/issues/4947 

### Cloud Provider(s):

gitrepo

### Environment:

spinnaker 1.28.1 clouddriver 5.76.2

### Description:

When the git repo contains a symlink, there is a chance the clouddriver throws an exception when trying to clean it up.

<details><summary>stack trace</summary>

```
java.io.IOException: Unable to delete directory /tmp/gitrepos/61273046364ed74b07a3d450a519abee22ed5f9ade99a891d8febc091a7dd44f.
        at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1127) ~[commons-io-2.7.jar:2.7]
        at com.netflix.spinnaker.clouddriver.artifacts.gitRepo.GitRepoArtifactCredentials.getLockedInputStream(GitRepoArtifactCredentials.java:100) ~[clouddriver-artifacts-5.76.2.jar:5.76.2]
        at com.netflix.spinnaker.clouddriver.artifacts.gitRepo.GitRepoArtifactCredentials.download(GitRepoArtifactCredentials.java:69) ~[clouddriver-artifacts-5.76.2.jar:5.76.2]
        at com.netflix.spinnaker.clouddriver.artifacts.ArtifactDownloader.download(ArtifactDownloader.java:37) ~[clouddriver-artifacts-5.76.2.jar:5.76.2]
        at com.netflix.spinnaker.clouddriver.controllers.ArtifactController.lambda$fetch$0(ArtifactController.java:69) ~[clouddriver-web-5.76.2.jar:5.76.2]
        at org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBodyReturnValueHandler$StreamingResponseBodyTask.call(StreamingResponseBodyReturnValueHandler.java:111) ~[spring-webmvc-5.2.12.RELEASE.jar:5.2.12.RELEASE]
        at org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBodyReturnValueHandler$StreamingResponseBodyTask.call(StreamingResponseBodyReturnValueHandler.java:98) ~[spring-webmvc-5.2.12.RELEASE.jar:5.2.12.RELEASE]
        at org.springframework.web.context.request.async.WebAsyncManager.lambda$startCallableProcessing$4(WebAsyncManager.java:337) ~[spring-web-5.2.12.RELEASE.jar:5.2.12.RELEASE]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:829) ~[na:na]
```

</details>

### Additional Details:

The root cause of it is that the clouddriver is using an old version of apache `commons-io` (2.7.0), which has the bug. It has been fixed by [this commit](https://github.com/apache/commons-io/commit/8fe59e41e8f092b33923bcb9a281fbc228a2bf8f) , in particular [these two lines](https://github.com/apache/commons-io/blob/rel/commons-io-2.8.0/src/main/java/org/apache/commons/io/file/DeletingPathVisitor.java#L137-L138). Upgrade it to 2.8.0 will have this issue fixed

**Comment by jasonmcintosh** [comment]
SO this has come up a few other places.  There's a feeling that there's another place where this is an issue as well, but a bump to commons-io is probably a good start (and maybe this WOULD fix the symlink issue)!  
https://github.com/spinnaker/kork/blob/ab04aa91ae6f7ab2dacb41c98db56bfa6fb0123e/spinnaker-dependencies/spinnaker-dependencies.gradle#L127
is where that's set.  Would welcome a PR to bump this and see if it'd help.  There's another issue on symlinks https://issues.apache.org/jira/browse/IO-751 that was hoping would be released to also fix some symlink issues.

The current solution is to turn on caching & set an infinite TTL & large space to cache the resources to get around this.  It's NOT ideal, and last I'd heard were waiting on an updated

**Comment by ovidiupopa07** [comment]
I will take a look and hopefully fix it

**Comment by weitie-okta** [comment]
Raised a PR with my personal account ^

**Comment by ovidiupopa07** [comment]
> Raised a PR with my personal account ^ 

And the exception is no longer thrown in clouddriver? I am asking because I use the 2.11.0 in clouddriver and the issue is still there

**Comment by tiewei** [comment]
> > Raised a PR with my personal account ^
> 
> And the exception is no longer thrown in clouddriver? I am asking because I use the 2.11.0 in clouddriver and the issue is still there

Do you have the stack trace? Is it the same as mine?

**Comment by ovidiupopa07** [comment]
It is the same. I can attach it on Monday, but it is the same exception> > > Raised a PR with my personal account ^
> 
> > 
> 
> > And the exception is no longer thrown in clouddriver? I am asking because I use the 2.11.0 in clouddriver and the issue is still there
> 
> 
> 
> Do you have the stack trace? Is it the same as mine?

**Comment by tiewei** [comment]
@ovidiupopa07  yes I have tested it with my use case, and it works as expected now (I tested with 2.8 when submitting the PR, and just tested with 2.11.0 today).
Our use case is during the kustomize stage it failed to cleanup the repo afterward, do you mind describing the use case if you still see the same error?

**Comment by ovidiupopa07** [comment]
hi @tiewei. I have a bake stage that clones a GitHub repo that has symlinks inside
![image](https://user-images.githubusercontent.com/105648914/202994772-bbea1e2a-eccd-465e-b273-00d0e47b5310.png)

Spinnaker: 1.29
Clouddriver: latest version from master + commons-io version bump

The Bake stage is failing after the version bump
```
java.io.IOException: Unable to delete directory /var/folders/27/pd06pk_52dd2sk4cz77b4phr0000gq/T/gitrepos/1731880f67c40628aa534a7755a0a36d1a9ddf837b52e88c60ef175a4999e1e3.
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1127)
	at com.netflix.spinnaker.clouddriver.artifacts.gitRepo.GitRepoArtifactCredentials.getLockedInputStream(GitRepoArtifactCredentials.java:100)
	at com.netflix.spinnaker.clouddriver.artifacts.gitRepo.GitRepoArtifactCredentials.download(GitRepoArtifactCredentials.java:69)
	at com.netflix.spinnaker.clouddriver.artifacts.ArtifactDownloader.download(ArtifactDownloader.java:37)
	at com.netflix.spinnaker.clouddriver.controllers.ArtifactController.lambda$fetch$0(ArtifactController.java:69)
	at org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBodyReturnValueHandler$StreamingResponseBodyTask.call(StreamingResponseBodyReturnValueHandler.java:111)
	at org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBodyReturnValueHandler$StreamingResponseBodyTask.call(StreamingResponseBodyReturnValueHandler.java:98)
	at org.springframework.web.context.request.async.WebAsyncManager.lambda$startCallableProcessing$4(WebAsyncManager.java:337)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
```
The way I fixed it is to use springframework utility method `FileSystemUtils.deleteRecursively(stagingPath.toFile());` instead of the commons-io `FileUtils.deleteDirectory(stagingPath.toFile());` It works fine with symlinks

**Comment by tiewei** [comment]
yea, the kork -> spinnaker-dependencies -> clouddriver builds. I'll open a backport PR too once this one is merged.

**Comment by tiewei** [comment]
@ovidiupopa07 The same pipeline config works fine with the 2.8.1 env I have (with the fix), would you confirm in your environment it's using the 2.11.0 version? These bug is more like the `commons-io` bug on `deleteDirectory` hence I don't think the spinnaker setup really matters, and a simple test code can be written to verify that. I believe the spring `deleteRecursively` works too. 

![image (1)](https://user-images.githubusercontent.com/4032269/203436339-90103127-a9cf-4cb4-8260-cc31afde3f75.png)
![image](https://user-images.githubusercontent.com/4032269/203436388-7c031919-6d9a-42cc-b5ac-e274a1b0826d.png)
![Screenshot](https://user-images.githubusercontent.com/4032269/203436405-9f88dcad-d0d8-4108-8f87-3bc7e9b49ec0.png)

**Comment by ovidiupopa07** [comment]
Hi @tiewe. Sorry for the late feedback, but I had a lot of other things on my plate. 
Everything works fine on my side too, we can close the issue. Thank you!

**Comment by tiewei** [comment]
no worries, thank you for confirming!
